### DISCUSSION TITLE
[SUGGESTION] Expansion-switch idea and historical poisonous ideas

### DISCUSSION THREAD

**Discussion by Helias** [azerothcore/azerothcore-wotlk#3912] [Ideas]
Hello everyone,

I want to discuss about these points:
- historical community cores/servers failures -> avoid splitting developers
- planning a cross-expansion configuration, at least for Classic, TBC and WOTLK

## Historical community cores/servers failures and avoid splitting developers
Personally, I used to work on the old database (UDB) of MaNGOS, I contributed to many servers working with MaNGOS and TrinityCore. I attended the migrations between those two cores.
I personally migrated from MaNGOS to TrinityCore and now I'm here on AzerothCore. I hope that my experience can make you think about this topic and avoid "poisonous ideas".
I mean, avoid the failures of MaNGOS, TrinityCore, server choices etc. I also suggest you to watch this video about communities and what we need to avoid as a OpenSource project https://www.youtube.com/watch?v=-F-3E8pyjFo

**Failures about MaNGOS**
MaNGOS after ArcEmu, was one of the most stable, used game server for WoW Classic and TBC expansions, but:
- they were a bit "slow" in merging new Pull Request (they didn't easily accept new changes)
- they were not much open to accept customizations
- they had DB, Scripts and Core communities split (UDBforums/YTDB, ScriptDev2 and MaNGOS forums and repositories)

MaNGOS couldn't avoid poisonous members / ideas that "drained" from their community a lot of developers that created new projects like TrinityCore, MaNGOSr2, etc.... instead of combining all development efforts into one project.

**Failures about TrinityCore**
TrinityCore was one of the most used and stable game servers for WOTLK. It evolved faster, full of fixes and developers, but:
- TrinityCore like MaNGOS, didn't avoid "needless infighting". For example the deletion of the 4.3.4 branch, which led to have some disappointed developers and the creation of TrinityCoreLegacy as a semi-separated project.
- TrinityCore was slow to react to new, innovative, ideas. Among the many examples, look at the [Yehonal's post about modularizing TrinityCore](https://community.trinitycore.org/topic/10536-trinitycore-as-a-true-mmo-framework/) which was basically ignored by most TC devs.

- after that SunwellWoW released the sources, TC failed to integrate SW features into the 3.3.5, saying that SW was just too hacky and ignored it.

So the community split their developers over and over...

**Failures about Servers**
- Any server that has a working and populated realm for 3.3.5a and tries to start a new realm for a new expansion, like cataclysm, after that, it eventually dies (both realms), unless it's a very BIG server.

That is because, having two realms splits the power of the developers and user base. If you don't double up the developers/game masters and users you can't double up your realm with a new expansion (already bugged from start probably).

## Recap

I think that from these experiences we can get that:
- open new projects split the human resources power (devs, community, etc..), so we should avoid that
- if people have new ideas, they should have a way to integrate it on the main existing project (AC modules is a good example), instead of creating a whole new project

That said, let's try further splitting and maybe even get more people involved to work at AzerothCore.

## Planning a cross-expansion configuration. At least for Classic, TBC and WOTLK

We currently support one expansion: WOTLK. We indeed have some folks here and there that would like to also work on different expansions. Creating new, separated projects would be bad for the reasons stated above.

Why don't we kill 2 birds with one stone? The pre-WOTLK content hasn't been rewritten in WOTLK (you can't say the same for Cataclysm, where the old world was heavily changed too).

So any pre-WOTLK content is pretty the same in AzerothCore too.

Some private servers, like PrimalWoW, had the idea of using the 3.3.5 client to start a progressive realm from classic and gradually unlock content until WOTLK.

Why if we, as AzerothCore, provided an easy way (through config) to quickly toggle expansions and contents ?

For example, if you want to open a TBC realm, you install AzerothCore and configure it to only provide TBC contents (i.e. max level is 70, no death knight, no access to Northrend, no WOTLK instances and items, etc...). The server administrator can choose to keep this configuration forever or change it one day to unlock WOTLK contents too.

The same can be done with Classic.

CONS:
- it wouldn't be **exaclty** playing TBC/Classic, but pretty close

PROS:
- user has access to some WOTLK features in Classic/TBC mode too (which can be optionally disabled)
- user has access to all WOTLK addons (there are many more than tbc/classic)
- server administrator has the ability to switch from classic to tbc and from tbc to wotlk, offering a truly progressive gaming experience just changing a conf value (without migrating DB or other time-consuming tasks)
- all devs work on 1 project, having 1 repository. No project split.

The last point can provide further benefits:

- getting more devs involved in AC, that want to work on Classic and/or TBC
- same dev tooling
- same modules suite available for different expansions
- ability of opening classic/tbc servers with some features that are available in wotlk only (e.g. BGs, Arenas, etc...)

I want to underline

**Comment by spleee254** [comment]
Not only do I find this to be a stellar idea, It is what I am musing on working on currently. I have been in the wow EMU community since 2007 and too went from mangos to trinitycore development. After my Last contributions I took a break from wow dev. After returning I was disappointed with the path TC has taken and found azerothCore to be a stellar replacement that focuses on a better direction. I think something like this would be fairly simple to begin implementing, and with the right planning could be controlled through modules that could ultimately add or remove features and content from specific patches. This is how I was looking into going about it. this would for example make it possible to potentialy separate tbc spells from wotlk etc. The big process to me would be unifying the data from  the expansions to work with AC. I would like to see this idea progress.

**Comment by BarbzYHOOL** [comment]
I haven't read all **yet** but I post just to say that I like the idea. And I wouldn't support a different expansion emulator (like mangos, or even TC did). But this is good

**Comment by ghost** [comment]
I also wouldn't support a different expansion emulator, but if this can be handled by configuration / module it would be a great idea. Personally I'd always choose a classic restricted WotLK over a true classic emulation as I found classic to be too cumbersome. TBC and WotLK had many good ideas concerning gameplay.

**Comment by Viste** [comment]
> THE IDEA IS ABOUT USING WOTLK CLIENT FOR CLASSIC-TBC TOO

Most insane and idiotic shit what i even hear/read. You don't have same experience on WOTLK client if you want play TBC or vanila you only get sliced WOTLK

**Comment by ghost** [comment]
Viste has a point, vanilla and tbc would be much more enjoyable. it's supposed to be a tedious, endless grinding for gold and exp and reputation and items. Talents and graphics would look different aswell, worse i mean! I kind of agree since i would never play on a Vanilla(wotlk) server but that's me and my opinion is just as valid as Viste's... 
However I do believe Azerothcore to be the best wow emulation from the community to the learning opportunity. Since I only care about programming and contributing I honestly don't care what people use the core for as long as there's happy people! 

Also modules are a great idea indeed but I don't care about them... All I wanna see is Azerothcore beat Trinitycore in terms of quality and reputation since I only had bad experiences with people there. Everyone is too arrogant and angry and salty and whatnot negative..

Let's unite brothers and sisters and script some good content into the repository 💪
